config/regex.js

```diff
@@ -50,7 +50,8 @@ module.exports = {
     INTERNAL_TRIGGER: /^~([\w-]+)$/,
     // Stages can only be named with A-Z,a-z,0-9,-,_
     STAGE_NAME: /^[\w-]+$/,
-
+    // Stage trigger like ~stage@deploy or ~stage@stageName:jobName
+    STAGE_TRIGGER: /^~stage@([\w-]+)(?::([\w-]+))?$/,
```

Comment:
`~` in the regex should be optional to support `AND` case
```suggestion
    // Stage trigger like ~stage@deploy or ~stage@stageName:jobName or stage@deploy or stage@stageName:jobName
    STAGE_TRIGGER: /^~?stage@([\w-]+)(?::([\w-]+))?$/,
```

Reply:
I think we can revisit join case later

Reply:
Looks like changes in workflow-parser is contradicting the schema. 

Schema has been configured to allow only `OR` case (i.e. the stage job references in `requires` must begin with `~stage@`).

But, workflow-parser only considers stage job references in `requires` that begins with `stage@` and ignores `~stage@`. Also by convention references with prefix `~` are considered as `OR` case. But for stages references without `~` are considered for `OR` case.
https://github.com/screwdriver-cd/workflow-parser/pull/43/files#r1243954316
https://github.com/screwdriver-cd/workflow-parser/pull/43/files#r1243947120
[Test data](https://github.com/screwdriver-cd/workflow-parser/blob/0f241d8e32e662e6abcbf8221aa7be415f580cab/test/lib/getWorkflow.test.js#L87) also omits `~` prefix
```
           jobs: {
                foo: { requires: ['~commit'] },
                A: { requires: ['foo'] },
                B: { requires: ['foo', 'stage@other:teardown'] },
                C: { requires: ['stage@deploy:setup'], stage: { name: 'deploy', startFrom: true } },
                D: { requires: ['~C'], stage: { name: 'deploy' } },
                E: { requires: ['stage@test:setup'], stage: { name: 'test', startFrom: true } },
                main: { requires: ['stage@other:setup'], stage: { name: 'other', startFrom: true } },
                publish: { requires: ['main'], stage: { name: 'other' } }
            }
```

Reply:
Yeah I think that's why I had the `^(?:~)?(stage@)` regex originally before you asked me to change it. I can have it only allow ~ prefix to keep it consistent. What do you think?

Reply:
> Yeah I think that's why I had the ^(?:~)?(stage@) regex originally before you asked me to change it.

Could you point me to the comment which suggested to update the regex `^(?:~)?(stage@)` to always expect the prefix `~`.

Reply:
https://github.com/screwdriver-cd/workflow-parser/pull/43#discussion_r1243947120

Reply:
https://github.com/screwdriver-cd/workflow-parser/pull/43#discussion_r1243947120
That code/comment has no reference to the said original regex `^(?:~)?(stage@)`.

That comment just suggests `~stage@` should be considered as `OR` case (this is the convention already being followed through out for non stage jobs) instead of `stage@` being considered as `OR` case stage jobs.
`stage@` by convention should be considered as `AND` case.

Reply:
sure

Reply:
For now we can just have OR support and we will revisit supporting AND for stages after we get this first pass for stages working.